run_iwyu_tests.py

```diff
@@ -87,26 +85,56 @@ def _AddTestFunctions(cls):
       while hasattr(cls, test_name):   # already have a class with that name
         test_name += '2'               # just append a suffix :-)
 
-      setattr(cls, test_name, _GetTestBody(filename))
+      setattr(cls, test_name, lambda x, f=filename: TestIwyuOnRelativeFile(f))
+      test_files[test_name] = filename
+
+    setattr(cls, 'test_files', test_files)
 
     return cls
 
   return _AddTestFunctions
 
 
-@GenerateTests(rootdir='tests/c', pattern='*.c')
-class c(unittest.TestCase):
-  pass
+def EnumerateLoadedTests():
+  for suite in unittest.defaultTestLoader.loadTestsFromModule(sys.modules[__name__]):
+    for test in suite:
+      yield (test.__class__, test._testMethodName)
+
+
+def PrintLoadedTests():
+  for (cls, test) in EnumerateLoadedTests():
+    print('%s.%s' % (cls.__name__, test))
 
 
-@GenerateTests(rootdir='tests/cxx', pattern='*.cc')
-class cxx(unittest.TestCase):
-  pass
+def PrintLoadedTestsAndFiles():
+  for (cls, test) in EnumerateLoadedTests():
+    print('%s.%s:%s' % (cls.__name__, test, cls.test_files[test]))
 
 
 if __name__ == '__main__':
   unittest_args, additional_args = Partition(sys.argv, '--')
   if additional_args:
     iwyu_test_util.SetIwyuPath(additional_args[0])
 
+  parser = argparse.ArgumentParser(add_help=False, usage=argparse.SUPPRESS)
+  group = parser.add_mutually_exclusive_group()
+  group.add_argument('--list', dest='list_tests', action='store_true')
+  group.add_argument('--list-test-files', action='store_true')
+  group.add_argument('--run-test-file')
+  (runner_args, _) = parser.parse_known_args(unittest_args)
+
+  if runner_args.run_test_file:
+    exit(TestIwyuOnRelativeFile(runner_args.run_test_file))
```

Comment:
I am not sure what exactly produces exit code `1`, but it works. I think the `raise AssertionError` should force-exit with code 1, otherwise something like `exit(None)` is being called.